apps/admin-portal/src/store/middleware/authenticate.ts

```diff
@@ -0,0 +1,149 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {
+    AuthenticateSessionUtil,
+    AuthenticateTokenKeys,
+    OPConfigurationUtil,
+    SignInUtil,
+    SignOutUtil
+} from "@wso2is/authentication";
+import { getAssociations, getProfileInfo } from "../../api";
+import { GlobalConfig, ServiceResourcesEndpoint } from "../../configs";
+import * as TokenConstants from "../../constants/token-constants";
+import { history } from "../../helpers";
+import { setProfileInfo, setSignIn, setSignOut } from "../actions";
+
+/**
+ * Handle user sign-out
+ *
+ * @param {object} state - AuthContext state object.
+ * @param {function} dispatch - State update `dispatch` react hook for AuthContext.
+ */
+export const handleSignIn = (state, dispatch) => {
+    const loginSuccessRedirect = () => {
+        const AuthenticationCallbackUrl = getAuthenticationCallbackUrl();
+        const location = ((!AuthenticationCallbackUrl)
+            || (AuthenticationCallbackUrl === APP_LOGIN_PATH)) ? APP_HOME_PATH : AuthenticationCallbackUrl;
+
+        history.push(location);
+    };
+
+    /**
+     * Get profile info and associations from the API
+     * to set the associations in the context.
+     */
+    const setProfileDetails = () => {
+        getProfileInfo()
+            .then((infoResponse) => {
+                getAssociations()
+                    .then((associationsResponse) => {
+                        dispatch(setProfileInfo({
+                            ...infoResponse,
+                            associations: associationsResponse
+                        }));
+                    });
+            });
+    };
+
+    const sendSignInRequest = () => {
+        const requestParams = {
+            clientHost: GlobalConfig.clientHost,
+            clientId: GlobalConfig.clientID,
+            clientSecret: null,
+            enablePKCE: true,
+            redirectUri: GlobalConfig.loginCallbackUrl,
+            scope: [ TokenConstants.LOGIN_SCOPE, TokenConstants.INTERNAL_IDENTITY_MGT.INTERNAL_IDENTITY_MGT_VIEW,
+                TokenConstants.INTERNAL_IDENTITY_MGT.INTERNAL_IDENTITY_MGT_UPDATE,
+                TokenConstants.INTERNAL_IDENTITY_MGT.INTERNAL_IDENTITY_MGT_DELETE,
+                TokenConstants.INTERNAL_IDENTITY_MGT.INTERNAL_IDENTITY_MGT_CREATE],
+        };
+        if (SignInUtil.hasAuthorizationCode()) {
+            SignInUtil.sendTokenRequest(requestParams)
+                .then((response) => {
+                    AuthenticateSessionUtil.initUserSession(response,
+                        SignInUtil.getAuthenticatedUser(response.idToken));
+                    dispatch(setSignIn());
+                    setProfileDetails();
+                    loginSuccessRedirect();
+                }).catch((error) => {
+                    throw error;
```

Comment:
Shouldn't we handle this error properly without throwing it? WDYT?

Reply:
This will be handled by the error handler in the Http module.